### PR TITLE
[ CNF-4622 ] - Trigger backup task and reconcile with the controller

### PR DESCRIPTION
**Pull request by sabbir-47**
Initial commit for triggering backup job
 -  added backup templates
 -  added backup implementation
 -  added conditions to launch jobs and monitoring the states 
 -  update the status back to operator
 -  modified reconcile logic
 -  updated rest api return when backup field is nil

### CODE REVIEW

File: controllers/backup.go [hunk 1]
```diff
@@ -104,7 +102,7 @@ func (r *ClusterGroupUpgradeReconciler) triggerBackup(ctx context.Context, clust
 
 // backupPreparing handles conditions in BackupStatePreparingToStart
 // returns: error
-func (r *ClusterGroupUpgradeReconciler) BackupPreparing(ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade,
+func (r *ClusterGroupUpgradeReconciler) backupPreparing(ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade,
```
Comment (danielmellado) [inline controllers/backup.go:105]: Are you switching this to non capital start as it's going to be used only as an internal helper?

**Comment by danielmellado** [comment]
Also, please whenever you're done with the commits, squash them, thanks!

File: controllers/backup.go (hunk 1 above)
Comment (sabbir-47) [reply controllers/backup.go:105]: yes, this function is exclusively used for the backup.

**Review by vitus133 — COMMENTED** [review]
Hey. just one minor comment, otherwise looks good.
I'm leaving it here for other folks to review.

File: controllers/backup.go [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+			return fmt.Errorf("cannot obtain the CGU cluster list: %s", err)
+		}
+		if clusterGroupUpgrade.Status.Backup == nil {
+			clusterGroupUpgrade.Status.Backup = &ranv1alpha1.BackupStatus{
+				Status:   make(map[string]string),
+				Clusters: clusters,
+			}
+		}
+
+		doneCondition := meta.FindStatusCondition(clusterGroupUpgrade.Status.Conditions, "BackupDone")
```
Comment (donpenney) [inline controllers/backup.go:42]: There are constants defined for backup states. Please add one for BackupDone, as it is used in a number of places throughout the file

File: controllers/backup.go [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+)
+
+// Backup states
+const (
+	BackupStatePreparingToStart = "PreparingToStart"
+	BackupStateStarting         = "Starting"
+	BackupStateActive           = "Active"
+	BackupStateSucceeded        = "Succeeded"
+	BackupStateTimeout          = "BackupTimeout"
+	BackupStateError            = "UnrecoverableError"
```
Comment (donpenney) [inline controllers/backup.go:21]: nit: If you're updating this file, sorting these alphabetically would be nice

File: controllers/precache.go [hunk 4]
```diff
@@ -296,11 +296,11 @@ func (r *ClusterGroupUpgradeReconciler) getPrecacheimagePullSpec(
 	return image, nil
 }
 
-// getPrecacheSpecTemplateData: Converts precaching payload spec to template data
-// returns: templateData (softwareSpec)
+// getPrecacheSpecprecacheTemplateData: Converts precaching payload spec to template data
```
[the remark this answers is not in the document]
Comment (sabbir-47) [reply controllers/precache.go]: Typo fixed with commit id ee44d60f05e97c9e24bae4295f068a15a16d2f04

File: controllers/precache.go [hunk 5]
```diff
@@ -216,7 +216,7 @@ func (r *ClusterGroupUpgradeReconciler) deployDependencies(
 	spec := r.getPrecacheSpecTemplateData(ctx, clusterGroupUpgrade)
 	spec.Cluster = cluster
 	msg := fmt.Sprintf("%v", spec)
-	r.Log.Info("[deployDependencies]", "getPrecacheSpecTemplateData",
+	r.Log.Info("[deployDependencies]", "getPrecacheSpecprecacheTemplateData",
```
Comment (donpenney) [inline controllers/precache.go:219]: Log needs to be updated

File: controllers/backup.go (hunk 3 above)
Comment (sabbir-47) [reply controllers/backup.go:21]: Kept it that way to show the flow of the status PreparingToStart -> Starting -> Active -> Final Status

Comment (donpenney) [reply controllers/backup.go:21]: Ok

File: controllers/backup.go (hunk 2 above)
Comment (sabbir-47) [reply controllers/backup.go]: Constant added @donpenney

File: controllers/backup.go [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+// backupPreparing handles conditions in BackupStatePreparingToStart
+// returns: error
+func (r *ClusterGroupUpgradeReconciler) backupPreparing(ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade,
+	cluster string) (string, error) {
+
+	currentState, nextState := BackupStatePreparingToStart, BackupStateStarting
+	r.Log.Info("[triggerBackup]", "currentState", currentState, "condition", "entry",
+		"cluster", cluster, "nextState", nextState)
+
+	// cleanup //
```
[the remark this answers is not in the document]
Comment (sabbir-47) [reply controllers/backup.go]: removed

File: controllers/backup.go [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+
+	r.Log.Info("[starting]", "conditions: ", condition)
+	switch condition {
+	case DependenciesNotPresent:
+
+		err := r.createResourcesFromTemplates(ctx, spec, backupDependenciesCreateTemplates)
+		if err != nil {
+			return currentState, err
+		}
+
```
[the remark this answers is not in the document]
Comment (sabbir-47) [reply controllers/backup.go:162]: done as suggested

File: controllers/managedClusterResources.go [hunk 8]
```diff
@@ -491,7 +533,26 @@ func (r *ClusterGroupUpgradeReconciler) getPrecacheJobTemplateData(
...
-// deployworkload deploys precaching workload on the spoke
+func (r *ClusterGroupUpgradeReconciler) getBackupJobTemplateData(clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade, clusterName string) *templateData {
+
+	rv := new(templateData)
+
+	rv.Cluster = clusterName
+	rv.JobTimeout = uint64(
+		clusterGroupUpgrade.Spec.RemediationStrategy.Timeout) * 60
+
+	// currently testing with hard-coded recovery image
```
[the remark this answers is not in the document]
Comment (sabbir-47) [reply controllers/managedClusterResources.go]: added //todo

File: controllers/templates/backup-templates.go [hunk 9]
```diff
@@ -70,18 +73,25 @@ spec:
...
+              target.workload.openshift.io/management: '{"effect":"PreferredDuringScheduling"}'
           spec:
             containers:
               -
                 args:
                   - launchBackup
                   - "--BackupPath"
                   - /var/recovery
-                image: {{ .WorkloadImage }}
+                image: quay.io/redhat_ztp/openshift-ai-image-backup:latest
```
[the remark this answers is not in the document]
Comment (sabbir-47) [reply controllers/templates/backup-templates.go]: Changed to recent image and added a //todo before constant declaration

File: controllers/precache.go (hunk 5 above)
Comment (sabbir-47) [reply controllers/precache.go]: updated

**Comment by donpenney** [comment]
No further comments from me. @vitus133, anything further?

**Pull request merged**